pkg/agent/nplagent/k8s/k8s_controller.go

```diff
@@ -0,0 +1,88 @@
+// +build !windows
+
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"time"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/nplagent/portcache"
+	"github.com/vmware-tanzu/antrea/pkg/util/env"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+const podResyncPeriod = 60 * time.Minute
+
+type Controller struct {
+	PortTable  *portcache.PortTable
+	KubeClient clientset.Interface
+}
+
+func NewNPLController(kubeClient clientset.Interface) *Controller {
+	return &Controller{KubeClient: kubeClient}
+}
+
+func (c *Controller) SetupEventHandlers(k8sinfo informers.SharedInformerFactory) {
+	nodeName, err := env.GetNodeName()
+	if err != nil {
+		klog.Warningf("Failed to get nodename, event handlers won't be started for NPL")
+		return
+	}
+	podEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			addPod := obj.(*corev1.Pod).DeepCopy()
```

Comment:
could we make deepcopy only when it's necessary, i.e. when we need to mutate the pod before making a API call? The deepcopy operation is not cheap.

Reply:
I didn't mean we don't make deepcopy but meant we make a deepcopy when we need to change the content of pod because there are cases in which pod won't be changed. I think `-race` won't complain on that.